src/components/SVGViewer/SVGViewer.tsx

```diff
@@ -450,6 +450,35 @@ export class SVGViewer extends Component<SvgViewerProps, SvgViewerState> {
     });
   };
 
+  addBulkClassesToMetadataContainer = (
+    metadataClassesAndConditions: {
+      condition: (metadataNode: Element) => boolean;
+      className: string;
+    }[]
+  ) => {
+    const gElements = Array.from(
+      this.svg.querySelectorAll('.metadata-container')
+    );
+
+    for (let i = 0; i < gElements.length; i++) {
+      const gElement = gElements[i];
+      const metadataElements = Array.from(
+        gElement.querySelectorAll('metadata')
+      );
+      const copy = [...metadataClassesAndConditions];
+      for (let j = 0; j < metadataElements.length - 1 && copy.length > 0; j++) {
+        const metadataElement = metadataElements[j];
+        for (let k = 0; k < copy.length - 1; k++) {
+          const conditionAndClass = copy[k];
+          if (conditionAndClass.condition(metadataElement)) {
+            gElement.classList.add(conditionAndClass.className);
+            copy.splice(k, 1);
+          }
+        }
+      }
+    }
+  };
```

Comment:
@ana-gheorghe if I understand things right, the performance improvement comes from the usage of the `copy` array saving calls to condition functions when all conditions have been met by a previous `metadataElement`, is that correct?

Assuming I got things right I've got a code style suggestion that you can ignore if dont like it and one change requests.

## Change request
`addBulkClassesToMetadataContainer` may be called more than once so we must clear the classList array first and add .metadata-container plus the classes that met the conditions.

## Suggestion (pseudo-code not tested)

Use [destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) and `for of` to add some clarity and save some unneeded lines

```
    for (const gElement of gElements) {
      const metadataElements = Array.from(
        gElement.querySelectorAll('metadata')
      );

      // reset classes in case this method is being called more than once for the same document
      gElement.className = 'metadata-container';

      const copy = [...metadataClassesAndConditions];
      for (const metadataElement of metadataElements) {
        // if another metadataElement has met all conditions
        // we dont need to check against the rest of metadataElements
        if (!copy.length) break;
        for (let k = 0; k < copy.length - 1; k++) {
          const { condition, className } = copy[k];
          if (condition(metadataElement)) {
            gElement.classList.add(className);
            copy.splice(k, 1);
          }
        }
      }
    }

Reply:
I believe PR should be a draft as I’ll take over the issue later. I hope this bulk thing will do some perf improvements but I have doubts to be honest that this makes huge difference regarding that weird issue we’ve seen.

Reply:
A huge difference, probably not, but it could help. I've spent many hours pinning down the problem and the bottlenecks. You also need to look into other parts, namely, the logic of the conditions themselves and how those are being applied to each metadata node. 

There are also other needed tweaks to improve "perceived performance" like opening up the modal and showing a loader first and then loading the SVG document inside instead of waiting for the SVG document to be downloaded and parsed before actually opening up the modal (with the UI frozen in the meantime).

I mentioned these and others things to Ana but if you're continuing with the task yourself, Im happy to show you too.

Reply:
Great, thanks for the info. Definitely like the idea about opening modal first to better visualize things.

Reply:
You can start by removing this line (https://github.com/cognitedata/infield/blob/034b945119d6a298e055b6c28f68ca0949dcd6da/src/components/SVGViewerWrapper/SVGViewerWrapper.tsx#L261) so that the SVGViewer takes care of the download itself.

Reply:
>I have doubts to be honest that this makes huge difference

My doubts were off the point here, I can see that works much faster on my machine now. So, good work @ana-gheorghe.

I however will play a bit more with it. I tried to apply you suggestions @franrull but I see issue with initial zoom in case when I do reset for className as you suggested. Not really sure how it's related though.

Reply:
> issue with initial zoom
Are you referring to zoom level or to centering the current asset.

If it is the latter, one of the classes is used to mark the current asset so that it is queried/found and the viewer is focused /centered on it. Maybe it is not find it?

Reply:
ok, so the problem was that current-asset class got reset and it shouldn't be. Fixed that. But I still have a problem that I can't open any document link from my mobile phone (not isafe), will need to look more into it before merging anything